File: Src/FUSE/Linux/x86-64/I-O_Engine/remount.c
/*
 * remount.c — cafs_remount(). Extracts the handful of crash-critical
 * fields this engine cares about from the Config Snapshot's TOML text
 * (already in ctx->config_snapshot from mount()) via a minimal,
 * targeted key scanner — not a general TOML parser. The Config
 * Snapshot is machine-generated, comment/blank-line-stripped text
 * (fs.info §5), so this is sufficient: track the current `[section]`,
 * match `key = value` lines only inside the sections we care about.
 *
 * Merges opts on top for the runtime-only knobs (readonly, direct_io,
 * sync_open, misaligned_action, destroy_flush — fs.info §5 explicitly
 * excludes these from the Config Snapshot). If the effective
 * direct_io/sync_open flags differ from how the fd is currently open,
 * closes and reopens it. Callable multiple times (that's the point —
 * see cafs_io.h's doc comment on the safe-mode-then-fix-then-final-
 * config workflow).
 *
 * SPDX-License-Identifier: BSD-3-Clause
 * Copyright (c) 2026, Imran Bin Gifary (System Delta)
 */
#include <fcntl.h>
#include <stdlib.h>
#include <string.h>
#include "cafs_io_internal.h"
#include "syscalls.h"

static int is_ws(char c) { return c == ' ' || c == '\t' || c == '\r'; }

static int is_section_line(const char *line, size_t linelen)
{
    size_t i = 0;
    while (i < linelen && is_ws(line[i])) i++;
    return (i < linelen && line[i] == '[');
}

static int key_match(const char *line, size_t linelen, const char *key,
                      const char **val, size_t *vallen)
{
    size_t i = 0;
    while (i < linelen && is_ws(line[i])) i++;
    size_t klen = strlen(key);
    if (linelen < i + klen) return 0;
    if (memcmp(line + i, key, klen) != 0) return 0;
    i += klen;
    while (i < linelen && is_ws(line[i])) i++;
    if (i >= linelen || line[i] != '=') return 0;
    i++;
    while (i < linelen && is_ws(line[i])) i++;
    size_t vstart = i, vend = linelen;
    for (size_t j = vstart; j < linelen; j++) {
        if (line[j] == '#') { vend = j; break; }
    }
    while (vend > vstart && is_ws(line[vend - 1])) vend--;
    if (vend > vstart && line[vstart] == '"' && line[vend - 1] == '"' && vend - vstart >= 2) {
        vstart++; vend--; /* strip surrounding quotes */
    }
    *val = line + vstart;
    *vallen = vend - vstart;
    return 1;
}

static uint32_t parse_uint(const char *val, size_t vallen)
{
    uint32_t n = 0;
    for (size_t i = 0; i < vallen && val[i] >= '0' && val[i] <= '9'; i++)
        n = n * 10 + (uint32_t)(val[i] - '0');
    return n;
}

static int strval_eq(const char *val, size_t vallen, const char *s)
{
    size_t slen = strlen(s);
    return vallen == slen && memcmp(val, s, slen) == 0;
}

static void scan_config_snapshot(const uint8_t *bytes, size_t cap, cafs_config_snapshot_t *out)
{
    memset(out, 0, sizeof(*out));
    size_t len = 0;
    while (len < cap && bytes[len] != '\0') len++; /* stop at the zero-padding */

    char section[64] = {0};
    size_t pos = 0;
    const char *text = (const char *)bytes;

    while (pos < len) {
        size_t eol = pos;
        while (eol < len && text[eol] != '\n') eol++;
        size_t linelen = eol - pos;
        const char *line = text + pos;

        if (is_section_line(line, linelen)) {
            size_t i = 0;
            while (i < linelen && is_ws(line[i])) i++;
            i++; /* skip '[' */
            size_t start = i;
            while (i < linelen && line[i] != ']') i++;
            size_t namelen = (i < linelen) ? (i - start) : 0;
            if (namelen >= sizeof(section)) namelen = sizeof(section) - 1;
            memcpy(section, line + start, namelen);
            section[namelen] = '\0';
        } else if (section[0] != '\0') {
            const char *val; size_t vallen;
            if (strcmp(section, "physical") == 0) {
                if (key_match(line, linelen, "block_size", &val, &vallen)) {
                    out->block_size = parse_uint(val, vallen);
                    out->fields_found |= CAFS_CFG_FOUND_BLOCK_SIZE;
                } else if (key_match(line, linelen, "pointer_checksum_algo", &val, &vallen)) {
                    if (strval_eq(val, vallen, "crc32c")) out->pointer_checksum_algo_id = CAFS_PTR_ALGO_CRC32C;
                    else if (strval_eq(val, vallen, "xxhash32")) out->pointer_checksum_algo_id = CAFS_PTR_ALGO_XXHASH32;
                    out->fields_found |= CAFS_CFG_FOUND_POINTER_ALGO;
                }
            } else if (strcmp(section, "integrity") == 0) {
                if (key_match(line, linelen, "data_checksum_algo", &val, &vallen)) {
                    if (strval_eq(val, vallen, "none")) out->data_checksum_algo_id = CAFS_DATA_ALGO_NONE;
                    else if (strval_eq(val, vallen, "crc32c")) out->data_checksum_algo_id = CAFS_DATA_ALGO_CRC32C;
                    else if (strval_eq(val, vallen, "xxhash64")) out->data_checksum_algo_id = CAFS_DATA_ALGO_XXHASH64;
                    else if (strval_eq(val, vallen, "blake3")) out->data_checksum_algo_id = CAFS_DATA_ALGO_BLAKE3;
                    out->fields_found |= CAFS_CFG_FOUND_DATA_ALGO;
                } else if (key_match(line, linelen, "verify_data", &val, &vallen)) {
                    out->verify_data = strval_eq(val, vallen, "true");
                    out->fields_found |= CAFS_CFG_FOUND_VERIFY_DATA;
                }
            }
        }
        pos = eol + 1;
    }
}

cafs_status_t cafs_remount(cafs_ctx_t *ctx, const cafs_mount_opts_t *opts,
                            cafs_config_snapshot_t *out_snapshot)
{
    if (!ctx) return CAFS_ERR_INVALID_ARG;
    if (!ctx->anchor_loaded) return CAFS_ERR_NOT_MOUNTED;

    pthread_mutex_lock(&ctx->lock);

    scan_config_snapshot(ctx->config_snapshot, sizeof(ctx->config_snapshot), &ctx->cfg);
    if (out_snapshot) *out_snapshot = ctx->cfg;

    cafs_mount_opts_t prev = ctx->opts;
    if (opts) ctx->opts = *opts;
    /* readonly can't be lifted by opts alone if the fd itself was
     * opened O_RDONLY at mount() — reopening below handles that. */

    int want_direct = ctx->opts.direct_io;
    int want_sync = ctx->opts.sync_open;
    int want_write_now = !ctx->opts.readonly;

    int need_reopen = (want_write_now != ctx->want_write) ||
                       (want_direct != prev.direct_io) ||
                       (want_sync != prev.sync_open);

    cafs_status_t st = CAFS_OK;
    if (need_reopen) {
        int flags = want_write_now ? O_RDWR : O_RDONLY;
        if (want_direct) flags |= O_DIRECT;
        if (want_sync) flags |= O_SYNC;

        int64_t newfd = cafs_sys_openat(CAFS_AT_FDCWD, ctx->device_path, flags, 0);
        if (newfd < 0) {
            if (ctx->opts.direct_io_required) {
                st = CAFS_ERR_UNSUPPORTED;
            } else {
                /* Fall back to the flags we already had; direct_io
                 * wasn't required, so this is a soft failure. */
                ctx->opts.direct_io = prev.direct_io;
                ctx->opts.sync_open = prev.sync_open;
                st = CAFS_OK;
            }
        } else {
            cafs_sys_close(ctx->fd);
            ctx->fd = (int)newfd;
            ctx->want_write = want_write_now;
        }
    }

    ctx->remounted = 1;
    pthread_mutex_unlock(&ctx->lock);
    return st;
}

File: Src/FUSE/Linux/x86-64/I-O_Engine/src/remount.rs
//! `CafsCtx::remount` — extracts the crash-critical fields this
//! engine needs from the Config Snapshot's TOML (a minimal targeted
//! key scanner, not a general TOML parser — the snapshot is machine-
//! generated, comment/blank-line-stripped text, regular enough that a
//! full parser wasn't worth building for four fields), then applies
//! caller-supplied [`MountOpts`] on top for the runtime-only knobs.
//! Callable multiple times: mount read-only (safe mode) -> remount
//! read-write so `check()` can persist a repair -> remount again with
//! the final on-disk config. Works because every anchor structure is
//! fixed at 4KB regardless of the filesystem's configured block size.
//!
//! SPDX-License-Identifier: BSD-3-Clause
//! Copyright (c) 2026, Imran Bin Gifary (System Delta)

use rustix::fs::{self, OFlags};

use crate::error::{CafsError, CafsResult};
use crate::layout::{PTR_ALGO_CRC32C, PTR_ALGO_XXHASH32};
use crate::{CafsCtx, ConfigSnapshot, MountOpts};

fn scan_config_snapshot(raw: &[u8]) -> ConfigSnapshot {
    let text = String::from_utf8_lossy(raw);
    let mut snap = ConfigSnapshot::default();
    let mut section = String::new();

    for line in text.lines() {
        let line = line.trim();
        if line.is_empty() {
            continue;
        }
        if let Some(inner) = line.strip_prefix('[').and_then(|s| s.strip_suffix(']')) {
            section = inner.to_string();
            continue;
        }
        let Some((key, value)) = line.split_once('=') else {
            continue;
        };
        let key = key.trim();
        let value = value.trim().trim_matches('"');

        match (section.as_str(), key) {
            ("physical", "block_size") => snap.block_size = value.parse().ok(),
            ("physical", "pointer_checksum_algo") => {
                snap.pointer_checksum_algo_id = match value {
                    "crc32c" => Some(PTR_ALGO_CRC32C),
                    "xxhash32" => Some(PTR_ALGO_XXHASH32),
                    _ => None,
                };
            }
            ("integrity", "data_checksum_algo") => {
                snap.data_checksum_algo_id = match value {
                    "none" => Some(crate::layout::DATA_ALGO_NONE),
                    "crc32c" => Some(crate::layout::DATA_ALGO_CRC32C),
                    "xxhash64" => Some(crate::layout::DATA_ALGO_XXHASH64),
                    "blake3" => Some(crate::layout::DATA_ALGO_BLAKE3),
                    _ => None,
                };
            }
            ("integrity", "verify_data") => snap.verify_data = value.parse().ok(),
            _ => {}
        }
    }

    snap
}

impl CafsCtx {
    pub fn remount(&mut self, opts: &MountOpts) -> CafsResult<ConfigSnapshot> {
        let snapshot = match &self.config_snapshot {
            Some(raw) => scan_config_snapshot(raw),
            None => ConfigSnapshot::default(),
        };

        if opts.readonly && self.want_write {
            // Downgrading write -> readonly in-process isn't
            // meaningful to enforce at the fd level (the fd stays
            // O_RDWR), so this is enforced in software: write_block/
            // confirm/check's repair path all check `opts.readonly`
            // before writing. want_write (fd-level) only ever
            // reflects what mount() originally opened with.
        } else if !opts.readonly && !self.want_write {
            // Caller wants to escalate to read-write, but the fd was
            // opened read-only at mount() — this can't be fixed with
            // fcntl (O_ACCMODE isn't changeable in place on Linux).
            // The caller needs a fresh cafs_mount(path, true, ...).
            return Err(CafsError::Unsupported);
        }

        let mut flags = fs::fcntl_getfl(&self.fd)?;
        flags.set(OFlags::DIRECT, opts.direct_io);
        flags.set(OFlags::SYNC, opts.sync_open);
        fs::fcntl_setfl(&self.fd, flags)?;

        if opts.direct_io && opts.direct_io_required {
            let now = fs::fcntl_getfl(&self.fd)?;
            if !now.contains(OFlags::DIRECT) {
                return Err(CafsError::Unsupported);
            }
        }

        self.opts = *opts;
        Ok(snapshot)
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn scans_all_four_fields() {
        let toml = b"[identity]\nname = \"test-volume\"\n\n[physical]\nblock_size = 4096\npointer_checksum_algo = \"crc32c\"\n\n[integrity]\ndata_checksum_algo = \"blake3\"\nverify_data = true\n";
        let snap = scan_config_snapshot(toml);
        assert_eq!(snap.block_size, Some(4096));
        assert_eq!(snap.pointer_checksum_algo_id, Some(PTR_ALGO_CRC32C));
        assert_eq!(snap.data_checksum_algo_id, Some(crate::layout::DATA_ALGO_BLAKE3));
        assert_eq!(snap.verify_data, Some(true));
    }

    #[test]
    fn ignores_unrelated_sections() {
        let toml = b"[mount_hints]\nreadonly = false\n\n[physical]\nblock_size = 8192\n";
        let snap = scan_config_snapshot(toml);
        assert_eq!(snap.block_size, Some(8192));
        assert_eq!(snap.verify_data, None);
    }
}
